src/modules/rlm_escape/rlm_escape.c

```diff
@@ -55,68 +63,86 @@ static char const hextab[] = "0123456789abcdef";
  *
  * @ingroup xlat_functions
  */
-static ssize_t escape_xlat(UNUSED TALLOC_CTX *ctx, char **out, size_t outlen,
-			   void const *mod_inst, UNUSED void const *xlat_inst,
-			   UNUSED request_t *request, char const *fmt)
+static xlat_action_t escape_xlat(TALLOC_CTX *ctx, fr_dcursor_t *out, request_t *request,
+				 void const *xlat_inst, UNUSED void *xlat_thread_inst,
+				 fr_value_box_list_t *in)
 {
-	rlm_escape_t const	*inst = mod_inst;
-	char const		*p = fmt;
-	char			*out_p = *out;
-	size_t			freespace = outlen;
-	size_t			len = talloc_array_length(inst->allowed_chars) - 1;
+	rlm_escape_t const	*inst;
+	void			*instance;
+	fr_value_box_t		*arg = fr_dlist_head(in);
+	char const		*p = arg->vb_strvalue;
+	char			*out_p;
+	size_t			reqlen = 0;
+	size_t			len;
+	fr_value_box_t		*vb;
+	fr_sbuff_t		sbuff;
+
+	memcpy(&instance, xlat_inst, sizeof(instance));
+	inst = talloc_get_type_abort(instance, rlm_escape_t);
+	len = talloc_array_length(inst->allowed_chars) - 1;
+
+	/*
+	 *	Calculate required size for escaped string
+	 */
+	while (p[0]) {
+		int chr_len = 1;
+		if (fr_utf8_strchr(&chr_len, inst->allowed_chars, len, p) == NULL) {
+			reqlen += chr_len * 3;	/* '=' + ([hex]{2}) */
+		} else {
+			reqlen += chr_len;
+		}
+		p += chr_len;
+	}
+
+	MEM(vb = fr_value_box_alloc_null(ctx));
+	if (fr_value_box_bstr_alloc(ctx, &out_p, vb, NULL, reqlen, false) < 0) {
```

Comment:
you can create a talloc sbuff using the box as the ctx, with an initial size of double the input string, then trim it at the end, and use the shallow copy functions to assign it to the box.  Probably faster and much less code.